### ISSUE TITLE
Zooeybot Disables Charge Attack on turn one after Enabling

### ISSUE THREAD

**Issue by justjulz29** [Masuzu/ZooeyBot#1224]
**Heya,**

I have an issue where on the first turn as soon as I have charge attack enabled, it disables it immediately afterwards. I've tried every possible solution; moving disable charge attack on turn 2 in my script, getting rid of it in turn 2, moving Enable Charge Attack to a different position... and it's still doing it. I'm utterly lost. Here is my Lua script I'm using for reference.

XXXXXXXXXXXXXXXXXXXXXXXX

 if turn == 1 then    
   character_1:UseSkill(1)
   character_1:UseSkill(2)
   character_1:UseSkill(3)
   character_1:UseSkill(4)
   character_2:UseSkill(1)
   character_2:UseSkill(2)
   character_3:UseSkill(1)
   character_3:OnPartyMember(1):UseSkill(2)
   character_4:UseSkill(1)
   character_4:UseSkill(2)
   character_4:UseSkill(3)
   Summon(6)
   EnableChargeAttack()
end 

if turn == 2 then 
   DisableChargeAttack()
   character_3:UseSkill(3)
   Summon(1)
end
   if character_1.charge_gauge >170 then
   EnableChargeAttack()
   character_3:UseSkill(3)
   Summon(1)
end

if turn == 3 then
   character_3:UseSkill(3) 
end

if turn == 4 then 
   Summon(5)
end 
  if character_1.charge_gauge <200 then
  DisableChargeAttack()
  Summon(5)
end

if turn == 5 then 
  EnableChargeAttack()
  Summon(4)
end 

if turn == 6 and character_4:HasStatusEffect("1111") then
  UseClarityHerbOnPartyMember(4)
  character_4:UseSkill(4)
  DisableChargeAttack()
  character_4:UseSkill(2)
  character_4:UseSkill(3)
  character_1:UseSkill(1)
  character_3:OnPartyMember(1):UseSkill(2)
end 
if enemy_1.charge_diamonds == 2 then 
  character_4:UseSkill(1) 
end

if turn == 7 then  
   character_1:UseSkill(4)
end
if enemy_1.charge_diamonds == 2 then 
  character_4:UseSkill(1)
  Summon(2)
end
if character_1.charge_gauge == 200 then
  EnableChargeAttack()
end 

if turn == 8 then
  EnableChargeAttack()
  Summon(6)
end
if character_1.charge_gauge >40 then 
  character_1:UseSkill(3)
end
if enemy_1.charge_diamonds == 2 then 
  character_4:UseSkill(1) 
  character_2:UseSkill(1)
  character_2:UseSkill(2)
  Summon(6)
end 

if turn == 9 then 
  DisableChargeAttack()
  UseGreenPotionOnPartyMember(1)
  character_1:UseSkill(2)
  character_2:UseSkill(3)
  character_4:UseSkill(2)
end
if character_1.charge_gauge == 200 then
  EnableChargeAttack()
  character_4:UseSkill(1)
  character_4:UseSkill(3)
end 

if turn == 10 and character_1.charge_gauge == 200 then 
  character_4:UseSkill(1)
  character_4:UseSkill(3)
end

**Comment by UminekoScarlet** [comment]
It might be this part:

```
if character_1.charge_gauge <200 then
	DisableChargeAttack()
	Summon(5)
end
```

It's not part of any of the turns, so it just checks if you got 200 charge gauge, then disables it. Since you use Summon(6) in turn 1 I expect that it's Bonito, so you will always have the 200 gauge on turn 1.

You can find it right after your turn = 4 part.

**Comment by UminekoScarlet** [comment]
Accidentally deleted my post just now instead of modifiying it.

The problem might be:
```
if character_1.charge_gauge <200 then
DisableChargeAttack()
Summon(5)
end
```
It's right after your turn == 4.

**Comment by justjulz29** [comment]
Oh bugger. How would I fix that then so I can still have that command work but not do what it is?

**Comment by UminekoScarlet** [comment]
Please tell me that turn 4 does, what class you are using and if the summon in turn 1 is Bonito.

**Comment by justjulz29** [comment]
I was able to work it out after much trial and error mate. Thanks for pointing out the flaw in the script!

**Comment by UminekoScarlet** [comment]
No worries.

You probably realized it by now, but the way you were using your turns made a lot of problems, since you wrote if statements that were supposed to be in turn x after the turn, so it tried doing it every single turn.

**Comment by zappky** [comment]
Justjulz, you just to better enclose your if statement. 

umineko has already pointed out that that line of code is outside of if statement which make it means. If character 1 charge bar is less than 200(this is almost always true). 
Disable charge attack and then try to summon 5th summon.
Since they are not enclosed in if statement checking on turn, this will be done on every single turn.

My advise is actually - dont micro manage what your unit is suppose to do on every turn,unless necessary such as for a big turn , Shiva summon turn,or dragon break turn.  Turn 1 is usually needed to be this detailed.

Firstly, the way you are doing now,it make your script a mess.
Secondly, even if you managed to really pinpoint your code to fix their every turn. There will be cases where your damage margin is off and the turn you scripted are off sync. 
lastly, it is just a pain to script this way.

You should view it in another way. View it by enemy hp threshold.
Since the bot will always try to use the skill as long as it is available you can simply it by  from which hp to which hp ,i want to **maintain** which skill/buff.

I usually put one big hp threshold check from 100 to 10, where i will put the skill in. 
10% and below are not scripted which means they will just be auto attacks.

I will the do smaller hp threshold check within it for further detailed script which is useful for much harder boss such as Akasha, Ubhl, and of course NM100 gw boss. 
There is no way you could script so in detail as the raid get harder. It will be a uphill challenge.

Those skill you want to keep them up, you could just put them outside of any if statement.

So in summary, only script using turn only when you want a very specific instruction for a very specific handling of buff.
Otherwise, make it more generic and flexible.

Hope this help you to script better.
